src/server/pkg/storage/chunk/reader.go

```diff
@@ -4,109 +4,183 @@ import (
+	if r.peek != nil {
+		dr := r.peek
+		r.peek = nil
+		return dr, nil
+	}
+	if len(r.dataRefs) == 0 {
+		return nil, io.EOF
+	}
+	dr := newDataReader(r.ctx, r.objC, r.dataRefs[0], r.prev)
+	r.dataRefs = r.dataRefs[1:]
+	r.prev = dr
+	return dr, nil
 }
 
-// Read reads from the byte stream produced by the set of DataRefs.
-func (r *Reader) Read(data []byte) (int, error) {
-	var totalRead int
-	defer func() {
-		r.len -= int64(totalRead)
-		r.bytesBeforeSplit += int64(totalRead)
-	}()
-	for len(data) > 0 {
-		n, err := r.r.Read(data)
-		data = data[n:]
-		totalRead += n
+// Iterate iterates over the data readers for the current data references
+// set in the reader.
+func (r *Reader) Iterate(f func(*DataReader) error) error {
+	for {
+		dr, err := r.Peek()
 		if err != nil {
-			if err := r.nextDataRef(); err != nil {
-				return totalRead, err
+			if err == io.EOF {
+				return nil
 			}
+			return err
+		}
+		if err := f(dr); err != nil {
+			return err
+		}
+		if _, err := r.Next(); err != nil {
+			return err
 		}
 	}
-	return totalRead, nil
+	return nil
 }
 
-func (r *Reader) nextDataRef() error {
-	// If all DataRefs have been read, then io.EOF.
-	if len(r.dataRefs) == 0 {
-		if r.f != nil {
-			dataRefs, err := r.f()
-			if err != nil {
-				return err
+// Get writes the concatenation of the data represented by the data references
+// set in the reader.
+// (bryce) probably should make a decision on whether this should be blocked for
+// a reader that already has been partially iterated.
+func (r *Reader) Get(w io.Writer) error {
+	return r.Iterate(func(dr *DataReader) error {
+		return dr.Get(w)
+	})
+}
+
+// DataReader is an abstraction that lazily reads data referenced by a data reference.
+// The seed is set to avoid re-downloading a chunk that is shared between this data reference
+// and the prior in a chain of data references.
+type DataReader struct {
+	ctx        context.Context
+	objC       obj.Client
+	dataRef    *DataRef
+	getChunkMu sync.Mutex
+	chunk      []byte
+	offset     int64
+	tags       []*Tag
+	seed       *DataReader
+}
+
+func newDataReader(ctx context.Context, objC obj.Client, dataRef *DataRef, seed *DataReader) *DataReader {
+	return &DataReader{
+		ctx:     ctx,
+		objC:    objC,
+		dataRef: dataRef,
+		offset:  dataRef.OffsetBytes,
+		tags:    dataRef.Tags,
+		seed:    seed,
+	}
+}
+
+// DataRef is the data reference associated with this data reader.
+func (dr *DataReader) DataRef() *DataRef {
+	return dr.dataRef
+}
+
+// Len is the length of the remaining data to be read.
+func (dr *DataReader) Len() int64 {
+	var size int64
+	for _, tag := range dr.tags {
+		size += tag.SizeBytes
+	}
+	return size
+}
+
+// Peek peeks ahead in the tags.
+func (dr *DataReader) Peek() (*Tag, error) {
+	if len(dr.tags) == 0 {
+		return nil, io.EOF
+	}
+	return dr.tags[0], nil
+}
+
+// Iterate iterates over the tags in the data reference and passes the tag and a reader for getting
+// the tagged content to the callback function.
```

Comment:
Could you describe what the `tagBound` parameter is in this comment as well? And I wondering if maybe there's a better name for it, `tagBound` doesn't mean much to me.

Reply:
Yeah, it essentially is an upper bound on the iteration. I'm not sure what a better name would be at the moment, but I will put some thought into it.

Reply:
Just calling it `tagUpperBound` or including that information in the comment would make it more clear. I'm a little confused what it means to have multiple `tagBound`s then if it's an upper bound, isn't only the smallest of the upperbounds relevant?

Reply:
It is an optional parameter, only the first element in the slice is relevant. I will change the name to `tagUpperBound` and update the comment.